### PR TITLE
Add a tuned plugin

### PR DESCRIPTION
**Pull request by evol262**

### CODE REVIEW

**Comment by stefwalter** [comment]
Nice :) Will test this out on Monday.

**Comment by stefwalter** [comment]
When running with git master I see the following issue:

```
TypeError: profile_selector.selectpicker is not a function
```

This probably means that we need to do something like:

```
diff --git a/pkg/tuned/tuned.js b/pkg/tuned/tuned.js
index 1afa27e..cdb7f12 100644
--- a/pkg/tuned/tuned.js
+++ b/pkg/tuned/tuned.js
@@ -24,7 +24,8 @@ define([
     "base1/cockpit",
     "translated!base1/po",
     "base1/mustache",
-    "base1/patterns"
+    "base1/patterns",
+    "base1/bootstrap-select",
 ], function($, cockpit, po, Mustache) {
     "use strict";
 ```

File: pkg/tuned/tuned.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            }
+            update_timeout = null;
+        }
+
+        function remove_notifications() {
+            $('div.container-fluid.alert').remove();
+        }
+
+        $('#tuned-start').on('click', function() {
+            var proc = cockpit.spawn(["systemctl", "restart", "tuned.service"]).
```
Comment (stefwalter) [inline pkg/tuned/tuned.js]: Recently @mvollmer has added a handy ```system/service``` wrapper javascript API to accomplish this task. You can see it being used in ```docker/containers.js```.

I think broadly we would do something like:

```
define([
...
    "system/service",
], function(..., service, ...) {
...
var tuned_service = service.proxy("tuned");
tuned_service.start().
      done(function () { ... }).
      fail(function (error) {
            show_curtain(cockpit.format(_("Failed to start Tuned: $0"), error));
      });
```

File: pkg/tuned/tuned.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        });
+
+        $('#tuned-change-button').on('click', function() {
+            wrapper('switch_profile', [$('#change-tuned').val()], function cb(result) {
+                console.warn(result[0]);
+                if (!result[0][0]) {
+                    $('#tuned-change-profile').dialog('failure', new Error("Problem applying tuned profile"));
+                    //TODO: log errors
+                }
+                $('#tuned-change-profile').hide();
```
Comment (stefwalter) [inline pkg/tuned/tuned.js]: This should use ```$('#tuned-change-profile').modal('hide');``` or the screen remains pretty gray after the dialog is gone.

File: pkg/tuned/tuned.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    tuned.register = function() {
+        $('#subscriptions-register-dialog').modal('show');
+    };
+
+    tuned.init = function(retrieve_current) {
+        if (retrieve_current === undefined)
+            retrieve_current = true;
+
+        var profile_selector = $('#subscription-register-url');
```
Comment (stefwalter) [inline pkg/tuned/tuned.js]: Looks like this is boilerplate that we don't need right?

File: pkg/tuned/tuned.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        return {
+            'get_profiles': get_profiles,
+            'verify_profile': verify_profile,
+            'recommend_profile': recommend_profile,
+            'switch_profile': switch_profile
+        };
+    }
+
+    tuned.register = function() {
+        $('#subscriptions-register-dialog').modal('show');
```
Comment (stefwalter) [inline pkg/tuned/tuned.js]: More boilerplate?

**Comment by stefwalter** [comment]
@andreasn This needs a proper design. Perhaps this would be a button on the system summary page which displays and allows changing the tuned profile? This would be similar to the realmd button there. WDYT?

**Comment by stefwalter** [comment]
Yay. I managed to use this to change my tuned profile :)

**Comment by stefwalter** [comment]
I've updated this trello card here: https://trello.com/c/eRlWwPcu/12-tuned-support

**Comment by stefwalter** [comment]
For the record here's how I got thish to work:

```
$ sudo yum install tuned
$ sudo systemctl start tuned
$ git fetch origin pull/3279/head
$ git checkout FETCH_HEAD
$ ln -snf $PWD/pkg/tuned ~/.local/share/cockpit/tuned
$ sudo systemctl restart cockpit
```

Now it shows up under Tools menu and should work.

**Comment by andreasn** [comment]
I get "Tuned" to show up under tools, but I get an "Oops" as soon as I go there.
This is the error output I get "TypeError: profile_selector.selectpicker is not a function"

**Comment by stefwalter** [comment]
> I get "Tuned" to show up under tools, but I get an "Oops" as soon as I go there. This is the error output I get "TypeError: profile_selector.selectpicker is not a function"

Ah yes, you need to add the line to tuned.js noted above.

**Comment by andreasn** [comment]
> Perhaps this would be a button on the system summary page which displays and allows changing the tuned profile?

I think that would make sense, yes, since the admin will be able to quickly see the result on the graphs to the right.
The label "Current Tuned Profile" sounds a bit jargon. Maybe better to call it "Perfomance profile" or something like that.

**Comment by evol262** [comment]
Updated a couple of these --

I'll redo it to use a service proxy and curtains tomorrow -- such rapid development is nice! That wasn't even there when I started on this...

**Comment by andreasn** [comment]
I've started a feature page for this https://github.com/cockpit-project/cockpit/wiki/Feature:Performance-tuning
Please check out the user stories and workflows and if they make sense.

**Comment by andreasn** [comment]
Is there any way, apart from looking at the conf-files to get an idea of what the different profiles do? Like a short summary? Or is it better to refer to the documentation for that?

**Comment by evol262** [comment]
Unfortunately, as far as I know, it ends up being documentation. They seem to have tried to name the profiles logically, but short of looking at the confs to see what sysctls it's setting/etc, I don't know of a better summary than the docs and performance tuning guides for Fedora/RHEL.

**Comment by mvollmer** [comment]
@evol262, we need to package this and add some integration tests for this, I am afraid.  I am happy to help.  Should I make a skeleton test?

**Comment by andreasn** [comment]
Here is my proposal for a design.
Move the button to the server page and have it pop up a dialog where None (tuned-adm off) and Recommended (tuned-adm recommend) are just items in the list.
More information about the different profiles would be hard, so it would be nice if there could be some helper text. Is there a good, standard way to link this against the correct documentation, or is it easier to skip that, and patch it in on the distros if needed later?

Mockup

**Comment by stefwalter** [comment]
> Move the button to the server page and have it pop up a dialog where None (tuned-adm off) and Recommended (tuned-adm recommend) are just items in the list.

This button should be disabled a button text if ```tuned``` is not available. The ```pkg/realmd``` 'Join Domain' button has similar behavior and implementation. I would suggest the tooltip text say "Cannot adjust performance because tuned is not available on this system"

```

@evol262 Are you interested in implementing the design @andreasn posted above? To clarify, Cockpit is design driven, and all significant features go through a workflow that looks like this: https://github.com/cockpit-project/cockpit/wiki/New-Features

So to take this beyond proof of concept we'd need to decide what to do:

 * Merge this proof of concept, but don't ship it in releases or packages, work on moving it to a designed feature in a separate pull request.
 * Implement the full feature, and then merge.

**Comment by andreasn** [comment]
> This button should be disabled a button text if tuned is not available. The pkg/realmd 'Join Domain' button has similar behavior and implementation. I would suggest the tooltip text say "Cannot adjust performance because tuned is not available on this system"

Sounds good to me!

**Comment by jeremyeder** [comment]
As the group responsible for tuned design and profile content, we want to sincerely thank you for doing this.  Wow guys.  Very very cool @evol262

**Comment by jeremyeder** [comment]
OK, read through -- yes, so tuned is not enabled by default on Fedora.  It is however enabled by default on RHEL7 and Atomic Host.

Very happy to see that the GUI will denote what profile was "recommended" by tuned, as that's the product setting it's per-variant default (simple regex matching in recommend.conf).

One very useful extension of this would be to surface whether or not the system is running the recommended profile on the base dashboard page.  So on this mockup https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/performance-profiles/performance-profiles.png you have a gray button that says "Balanced".  If you could somehow indicate at that level whether or not the applied profile is the recommended profile, that would be super helpful.  If the applied profile is not the one recommended by tuned, then you could write 'customized' or 'site-specific' or something like that.


@andreasn "More information about the different profiles would be hard" 

The best (not ideal) source of this is probably the "tuned-profiles" man page.  Is it possible to fish this information out of there?

**Comment by stefwalter** [comment]
> OK, read through -- yes, so tuned is not enabled by default on Fedora. It is however enabled by default on RHEL7 and Atomic Host.

We can have UI to turn it on if it's installed, but not enabled.

> @andreasn "More information about the different profiles would be hard" The best (not ideal) source of this is probably the "tuned-profiles" man page. Is it possible to fish this information out of there?

That would be pretty hard to display in context. Is there a possibility of a future enhancement to have a "description" retrievable via the DBus API. This description could be succinct information about the profile.

Adding such a "description" to the API wouldn't block this work, but once available we could display it in Cockpit.

**Comment by jeremyeder** [comment]
@andreasn @stefwalter ok, Jaroslav ( @yarda ) is going to extend the tuned API to allow Cockpit to query tuned profile descriptions.  We're going to embed descriptions in a new section of the tuned.conf files included with each profile, and the new tuned API will know how to parse that.

We'll be writing those and merging them into tuned upstream in the next couple days.

**Comment by jeremyeder** [comment]
Actually... @stefwalter what did you mean by "succinct" ... Is there a character limit you had in mind? I was hoping to be slightly verbose in the descriptions, perhaps 2 or 3 sentences each.  If you need a "short description", with say, and 80char limit, I can do that in addition?

**Comment by stefwalter** [comment]
> Actually... @stefwalter what did you mean by "succinct" ... Is there a character limit you had in mind? I was hoping to be slightly verbose in the descriptions, perhaps 2 or 3 sentences each. If you need a "short description", with say, and 80char limit, I can do that in addition?

I think 2 to 3 sentences sounds good. 

@andreasn What do you think? What did you have in mind with the design?

**Comment by yarda** [comment]
Or I can add to the API possibility to get summary (short text) and description (long text) similarly to the RPM.

**Comment by jeremyeder** [comment]
@yarda what would the config file look like in a case where we had both brief and verbose descriptions?  API assumes line 1 below [] is brief and line 2 --> next set of [] is long?

[description]
my short description
my longer description goes here

**Comment by yarda** [comment]
I currently think about e.g. the following syntax of the tuned.conf:
[main]
summary=SHORT_TEXT
description=LONG_TEXT

Multiple lines text in quotes, e.g:
[main]
description="LONG_TEXT_LINE1
LONG_TEXT_LINE2
LONG_TEXT_LINE3"

I think this would require the easiest change of the parser.

**Comment by andreasn** [comment]
> What do you think? What did you have in mind with the design?

I haven't quite figured out exactly where to put it, but I'll try to figure something out. One or two lines would be enough I think.

**Comment by andreasn** [comment]
> If you could somehow indicate at that level whether or not the applied profile is the recommended profile

Sure, I'll add that to the design.

**Comment by jeremyeder** [comment]
patch written

**Comment by stefwalter** [comment]
> Sure, I'll add that to the design.

The other thing that needs to go into the design is how the button is disabled, its tooltip, and what it says when tuned is not available.

**Comment by andreasn** [comment]
These are the different button states.
I'm not super-sure if not running a recommended profile should be consithered an error. If an admin change to network-throughput because he's running a file-server, is that really an error?
Does that make sense, or did I get recommended backwards?

buttonstates mockup

**Comment by jeremyeder** [comment]
Thanks for checking.  Not running an recommended profile is _not_ an error, because the only way to for that to occur is for someone to do it manually.  I was thinking it still should be noted "somehow" in the GUI that it's not the default.  I had suggested "site-specific" or "customized"?

**Comment by andreasn** [comment]
> Not running an recommended profile is not an error, because the only way to for that to occur is for someone to do it manually.

Ah, ok. I changed the mockup to use a (i) icon and show more info when you click on it. Some of the profile names use long names, so to save on space, I think this tooltip should do the trick.

**Comment by jeremyeder** [comment]
Looks good, yep. Also I sent the patch to add the profile descriptions yesterday to @yarda ...I assume that will show up in RHEL7.3 and F24.

**Comment by andreasn** [comment]
And here is the dialog with the profile descriptions.

dialog mockup

**Comment by yarda** [comment]
> Looks good, yep. Also I sent the patch to add the profile descriptions yesterday to @yarda ...I assume that will show up in RHEL7.3 and F24.

I am going to merge it Tuned upstream during next week.

**Comment by andreasn** [comment]
> I am going to merge it Tuned upstream during next week. 

That's great! Thanks for the quick response on this everyone!

**Comment by yarda** [comment]
The tuned-2.6.0 is in Fedora updates-testing. This version adds summary and description support. Summary is short text about profile purpose. Description is longer text which can be multiline (currently there is no description written for available profiles). This Tuned version extends its D-Bus API the following way:

- adds 'profile_info' method. It gets profile_name argument and it queries the requested profile for summary/description. If profile_name is empty string it queries the active/selected profile. It returns tuple of the following form (status, profile_name, summary, description), where status is boolean. If it is False, the query mechanism failed and the other fields don't contain valid data. If there is no summary or description available for the queried profile the appropriate string is empty.

- adds 'profiles2' method. It's extension of 'profiles' method. It returns list of tuples. Tuples are of the following form: (profile_name, profile_summary)

In Tuned the summary/description is set in profile configuration (tuned.conf) under the main section, i.e.:
```
[main]
summary=TEXT
description=TEXT
```
multiline text can be specified the following way:
```
[main]
description='''LINE1
LINE2'''
```
If there is no main section or summary/description is missing, Tuned treats it as empty strings.

**Comment by yarda** [comment]
It seems that new lines embedded in description are presented through D-Bus as '\n' strings (at least through d-feet).

**Comment by yarda** [comment]
tuned-adm command was also extended, 'tuned-adm list' now includes summary, and there is also 'tuned-adm profile_info [profile]' which shows summary and description, if no profile is given it queries the current profile.

**Pull request closed without merging**